### PR TITLE
CIRC-775 fix NPE, create test

### PR DESCRIPTION
**Pull request by roman-barannyk**
## Approach 
- fix NPE during Loan Policy processing with a fixed due date schedule;
- create tests;

Resolves [CIRC-775](https://issues.folio.org/browse/CIRC-775)

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [x] There are no breaking changes in this PR.

### CODE REVIEW

File: src/main/java/org/folio/circulation/domain/policy/LoanPolicy.java [hunk 1]
```diff
@@ -324,7 +324,8 @@ public LoanPolicy withDueDateSchedules(JsonObject fixedDueDateSchedules) {
   }
 
   LoanPolicy withAlternateRenewalSchedules(FixedDueDateSchedules renewalSchedules) {
-    return new LoanPolicy(representation, fixedDueDateSchedules, renewalSchedules);
+    return new LoanPolicy(
+      representation, fixedDueDateSchedules, renewalSchedules, ruleConditionsEntity);
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/domain/policy/LoanPolicy.java:328]:
```suggestion
    return new LoanPolicy(representation, fixedDueDateSchedules, renewalSchedules, 
      ruleConditionsEntity);
```

File: src/main/java/org/folio/circulation/domain/policy/LoanPolicy.java [hunk 2]
```diff
@@ -315,7 +315,7 @@ private boolean isProfile(JsonObject loansPolicy, String profileId) {
 
   LoanPolicy withDueDateSchedules(FixedDueDateSchedules loanSchedules) {
     return new LoanPolicy(representation, loanSchedules,
-      alternateRenewalFixedDueDateSchedules);
+      alternateRenewalFixedDueDateSchedules, ruleConditionsEntity);
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/domain/policy/LoanPolicy.java:318]: What is a `ruleConditionsEntity`?

Should it be required for all instances of `LoanPolicy`?

Comment (roman-barannyk) [reply src/main/java/org/folio/circulation/domain/policy/LoanPolicy.java:318]: ruleConditionsEntity is an object which we use for itemLimit checking. It contains 3 fields:   - boolean isItemTypePresent;
boolean isLoanTypePresent;
boolean isPatronGroupPresent;
That indicates which conditions were applied for matching circulation rule.

Comment (roman-barannyk) [reply src/main/java/org/folio/circulation/domain/policy/LoanPolicy.java:318]: > Should it be required for all instances of LoanPolicy?

Yes, it should. For this case with fixed due date schedule, new LoanPolicy was created without ruleConditionsEntity. That was a reason why we got NPE in ItemLimitValidator.

Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/domain/policy/LoanPolicy.java:318]: > > Should it be required for all instances of LoanPolicy?
> 
> Yes, it should. For this case with fixed due date schedule, new LoanPolicy was created without ruleConditionsEntity. That was a reason why we got NPE in ItemLimitValidator.

Please could it be made a required constructor parameter (by removing any overload that does not have it) and possibly disallowing null values for that parameter

Comment (roman-barannyk) [reply src/main/java/org/folio/circulation/domain/policy/LoanPolicy.java:309]: Thanks, removed

File: src/main/java/org/folio/circulation/domain/policy/LoanPolicy.java [hunk 3]
```diff
@@ -57,17 +57,8 @@
...
-
-  LoanPolicy(
-    JsonObject representation,
-    FixedDueDateSchedules fixedDueDateSchedules,
-    FixedDueDateSchedules alternateRenewalFixedDueDateSchedules) {
-    super(getProperty(representation, "id"), getProperty(representation, "name"));
-    this.representation = representation;
-    this.fixedDueDateSchedules = fixedDueDateSchedules;
-    this.alternateRenewalFixedDueDateSchedules = alternateRenewalFixedDueDateSchedules;
+      new NoFixedDueDateSchedules(),
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/domain/policy/LoanPolicy.java:60]: Can this be included on the previous line? I don't think the parameters need to be split one per line (that is likely a hangover from a style I used to use and probably should start moving away from)

File: src/test/java/api/loans/CheckOutByBarcodeTests.java [hunk 4]
```diff
@@ -1164,6 +1165,34 @@ public void cannotCheckOutWhenItemLimitIsReachedForBookMaterialTypeAndLoanTypeIs
...
+    IndividualResource videoTypeItem = itemsFixture.basedUponDunkirk();
+    IndividualResource steve = usersFixture.steve();
+
+    checkOutFixture.checkOutByBarcode(firstBookTypeItem, steve);
+    firstBookTypeItem = itemsClient.get(firstBookTypeItem);
+    assertThat(firstBookTypeItem, hasItemStatus(CHECKED_OUT));
+
+    Response response = checkOutFixture.attemptCheckOutByBarcode(secondBookTypeItem, steve);
+    assertThat(response.getJson(), hasErrorWith(allOf(
+      hasMessage("Patron has reached maximum limit of 1 items for material type"))));
```
Comment (marcjohnson-kint) [inline src/test/java/api/loans/CheckOutByBarcodeTests.java:1187]: If only one aspect of the error is being checked, I don't think the `allOf` matcher is needed

```suggestion
    assertThat(response.getJson(), hasErrorWith(hasMessage("Patron has reached maximum limit of 1 items for material type"))));
```

File: src/test/java/api/loans/CheckOutByBarcodeTests.java [hunk 5]
```diff
@@ -1164,6 +1165,34 @@ public void cannotCheckOutWhenItemLimitIsReachedForBookMaterialTypeAndLoanTypeIs
...
+    secondBookTypeItem = itemsClient.get(secondBookTypeItem);
+    assertThat(secondBookTypeItem, hasItemStatus(AVAILABLE));
```
Comment (marcjohnson-kint) [inline src/test/java/api/loans/CheckOutByBarcodeTests.java:1189]: I might include some whitespace above these statements to separate them from the previous section and assertion.

**Pull request merged**